activerecord/lib/active_record/associations/has_many_through_association.rb

```diff
@@ -53,16 +53,24 @@ def concat_records(records)
         #
         # However, after insert_record has been called, the cache is cleared in
         # order to allow multiple instances of the same record in an association.
-        def build_through_record(record)
-          @through_records[record] ||= begin
-            ensure_mutable
+        def through_records(record)
+          @through_records[record] ||= build_through_record(record)
+        end
 
-            attributes = through_scope_attributes
-            attributes[source_reflection.name] = record
-            attributes[source_reflection.foreign_type] = options[:source_type] if options[:source_type]
+        def build_through_record(record)
+          ensure_mutable
 
-            through_association.build(attributes)
+          if (inverse = source_reflection.inverse_of) && inverse.has_one? &&
```

Comment:
Could you move the assignment outside the `if` statement?
The if statement is pretty complex and we assign outside the `if` in the same file on other places.
https://github.com/rails/rails/blob/f5b74ff2ac298c11f06cd86120fc9af78074a68d/activerecord/lib/active_record/associations/has_many_through_association.rb#L96

```suggestion
          inverse = source_reflection.inverse_of
          if inverse && inverse.has_one? &&
```